### PR TITLE
Fix for FullStatus gRPC connection pooling

### PR DESCRIPTION
**Pull request by GuptaManan100**
## Description

Fixes a bug in `TabletManagerClient`'s pool dialer. In:

https://github.com/vitessio/vitess/blob/8833092c2e48469e981ca4005f43f334a2305dd9/go/vt/vttablet/grpctmclient/client.go#L163-L170

`rpcClientMap` caches tablet connections indefinitely based on address. If a tablet goes down, it does not get evicted from the map. If a new tablet then bootstraps with the same address, then `rpcClientMap` makes us continuously try connecting to that tablet, even if it's on another cluster.

The solution is to evict a tablet/connection from the map upon connection/gRPC error.

However, since said error is seen long after we've returned the cached client / dialer, the user of the client needs to proactively invalidate the cached connection. To that effect, we now return an `invalidator()` function that can invalidate a cached connection based on the returned error.

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vttablet/grpctmclient/client.go [hunk 1]
```diff
@@ -185,7 +186,19 @@ func (client *grpcClient) dialPool(ctx context.Context, tablet *topodatapb.Table
...
 	c <- result
-	return result.client, nil
+	invalidator := func(err error) {
+		if err == nil {
+			return
+		}
+		if _, ok := status.FromError(err); !ok {
+			// Not a gRPC error
+			return
+		}
```
Comment (GuptaManan100) [inline go/vt/vttablet/grpctmclient/client.go:196]: @harshit-gangal Would some error codes from MySQL also pass this check and end up causing us to invalidate the grpc connection?

File: go/vt/vttablet/grpctmclient/client.go [hunk 2]
```diff
@@ -185,7 +186,19 @@ func (client *grpcClient) dialPool(ctx context.Context, tablet *topodatapb.Table
...
+		client.mu.Lock()
+		defer client.mu.Unlock()
+		delete(client.rpcClientMap, addr)
```
Comment (GuptaManan100) [inline go/vt/vttablet/grpctmclient/client.go:199]: We should go over the map and close all the grpc connections - 
```go
for tm := range client.rpcClientMap[addr] {
	tm.cc.Close()
}
```

**Review by GuptaManan100 — COMMENTED** [review]
Rest LGTM

File: go/vt/vttablet/grpctmclient/client.go (hunk 2 above)
Comment (shlomi-noach) [reply go/vt/vttablet/grpctmclient/client.go:199]: Done.

File: go/vt/vttablet/grpctmclient/client.go [hunk 3]
```diff
@@ -1120,6 +1140,9 @@ func (client *Client) CheckThrottler(ctx context.Context, tablet *topodatapb.Tab
 
 	response, err := c.CheckThrottler(ctx, req)
 	if err != nil {
+		if invalidator != nil {
+			invalidator(err)
+		}
 		return nil, err
 	}
```
Comment (harshit-gangal) [inline go/vt/vttablet/grpctmclient/client.go:1147]: nit: can remove passing `error` to `invalidator`

**Comment by shlomi-noach** [comment]
Converted this PR to draft, as we explore alternative approaches: #15541, #15559

**Comment by shlomi-noach** [comment]
A different implementation in https://github.com/vitessio/vitess/pull/15562 .

**Pull request closed without merging**